archive/imo/imo2019_q2.lean

```diff
@@ -0,0 +1,576 @@
+
+section oriented
+
+variables [module.oriented ℝ V (fin 2)]
+
+lemma two_zsmul_oangle_QPA₂_eq_two_zsmul_oangle_BAA₂ :
+  (2 : ℤ) • ∡ cfg.Q cfg.P cfg.A₂ = (2 : ℤ) • ∡ cfg.B cfg.A cfg.A₂ :=
+begin
+  refine two_zsmul_oangle_of_parallel cfg.QP_parallel_BA _,
+  convert affine_subspace.parallel.refl _ using 1,
+  rw [cfg.collinear_PAA₁A₂.affine_span_eq_of_ne
+        (set.mem_insert_of_mem _ (set.mem_insert_of_mem _ (set.mem_insert_of_mem _
+          (set.mem_singleton _))))
+        (set.mem_insert_of_mem _ (set.mem_insert _ _)) cfg.A₂_ne_A,
+      cfg.collinear_PAA₁A₂.affine_span_eq_of_ne
+        (set.mem_insert_of_mem _ (set.mem_insert_of_mem _ (set.mem_insert_of_mem _
+          (set.mem_singleton _))))
+        (set.mem_insert _ _) cfg.A₂_ne_P]
+end
+
+end oriented
+
+lemma not_collinear_QPA₂ : ¬ collinear ℝ ({cfg.Q, cfg.P, cfg.A₂} : set Pt) :=
+begin
+  haveI := some_orientation V,
+  rw [collinear_iff_of_two_zsmul_oangle_eq cfg.two_zsmul_oangle_QPA₂_eq_two_zsmul_oangle_BAA₂,
+      ←affine_independent_iff_not_collinear_set],
+  have h : cospherical ({cfg.B, cfg.A, cfg.A₂} : set Pt),
+  { refine cfg.triangle_ABC.circumsphere.cospherical.subset _,
+    simp [set.insert_subset, cfg.A_mem_circumsphere, cfg.B_mem_circumsphere,
+          cfg.A₂_mem_circumsphere] },
+  exact h.affine_independent_of_ne cfg.A_ne_B.symm cfg.A₂_ne_B.symm cfg.A₂_ne_A.symm
+end
+
+lemma Q₁_ne_A₂ : cfg.Q₁ ≠ cfg.A₂ :=
+begin
+  intro h,
+  have h₁ := cfg.sbtw_Q_A₁_Q₁,
+  rw h at h₁,
+  refine cfg.not_collinear_QPA₂ _,
+  have hA₂ := cfg.sbtw_A_A₁_A₂.right_mem_affine_span,
+  have hA₂A₁ : line[ℝ, cfg.A₂, cfg.A₁] ≤ line[ℝ, cfg.A, cfg.A₁] :=
+    affine_span_pair_le_of_left_mem hA₂,
+  have hQ : cfg.Q ∈ line[ℝ, cfg.A, cfg.A₁],
+  { rw affine_subspace.le_def' at hA₂A₁,
+    exact hA₂A₁ _ h₁.left_mem_affine_span },
+  exact collinear_triple_of_mem_affine_span_pair hQ cfg.wbtw_A_P_A₁.mem_affine_span hA₂
+end
+
+lemma affine_independent_QPA₂ : affine_independent ℝ ![cfg.Q, cfg.P, cfg.A₂] :=
+affine_independent_iff_not_collinear_set.2 cfg.not_collinear_QPA₂
+
+lemma affine_independent_PQB₂ : affine_independent ℝ ![cfg.P, cfg.Q, cfg.B₂] :=
+by { rw ←symm_A₂, exact cfg.symm.affine_independent_QPA₂ }
+
+/-- `QPA₂` as a `triangle`. -/
+def triangle_QPA₂ : triangle ℝ Pt := ⟨_, cfg.affine_independent_QPA₂⟩
+
+/-- `PQB₂` as a `triangle`. -/
+def triangle_PQB₂ : triangle ℝ Pt := ⟨_, cfg.affine_independent_PQB₂⟩
+
+lemma symm_triangle_QPA₂ : cfg.symm.triangle_QPA₂ = cfg.triangle_PQB₂ :=
+by { simp_rw [triangle_PQB₂, ←symm_A₂], ext i, fin_cases i; refl }
+
+/-- `ω` is the circle containing `Q`, `P` and `A₂`, which will be shown also to contain `B₂`,
+`P₁` and `Q₁`. -/
+def ω : sphere Pt := cfg.triangle_QPA₂.circumsphere
+
+lemma P_mem_ω : cfg.P ∈ cfg.ω := cfg.triangle_QPA₂.mem_circumsphere 1
+
+lemma Q_mem_ω : cfg.Q ∈ cfg.ω := cfg.triangle_QPA₂.mem_circumsphere 0
+
+section oriented
+
+variables [module.oriented ℝ V (fin 2)]
+
+lemma two_zsmul_oangle_QPA₂_eq_two_zsmul_oangle_QB₂A₂ :
```

Comment:
I think Oliver has brought this up in the past, but I guess it will confuse readers when the model solution says "we show angle QPA_2 equals angle QB_2A_2" and the Lean solution has this "strange" multiplication by two because we don't have oriented angles mod pi.

Reply:
I also remarked on this in the module doc string (referencing the `geometry.euclidean.angle.oriented.basic` doc string that also discusses this issue).

Reply:
Indeed, I speculated about introducing a type for angles mod `π` [here](https://github.com/leanprover-community/mathlib/pull/17525#issuecomment-1330799041) but I tend to agree with @jsm28 that it's probably not worth the trouble.

It's a bit disappointing we don't have a really neat solution for this, but I certainly don't think such matters should delay this PR.